modules/dnn/src/layers/layer_norm.cpp

```diff
@@ -0,0 +1,172 @@
+        {
+            CV_Assert(src->isContinuous());
+            CV_Assert(dst->isContinuous());
+            CV_CheckType(src->type(), CV_32F, "DNN/LayerNorm: only support float32");
+            CV_Assert(src->type() == dst->type());
+
+            LayerNormInvoker p;
+
+            p.src = src;
+            p.scale = scale;
+            p.bias = b;
+            p.dst = dst;
+
+            p.epsilon = epsilon;
+            p.nstripes = nstripes;
+
+            auto dstShape = shape(*dst);
+            p.total = std::accumulate(dstShape.begin(), dstShape.begin() + axis, 1, std::multiplies<int>());
+            p.stripeSize = (p.total + nstripes - 1) / nstripes;
+            p.normSize = std::accumulate(dstShape.begin() + axis, dstShape.end(), 1, std::multiplies<int>());
+
+            parallel_for_(Range(0, nstripes), p, nstripes);
+        }
+
+        void operator()(const Range& r) const CV_OVERRIDE
+        {
+            int stripeStart = r.start * stripeSize;
+            int stripeEnd = std::min(r.end * stripeSize, total);
+
+            float *dstData = (float *)dst->data;
+            const float *srcData = (const float *)src->data;
+            const float *scaleData = (const float *)scale->data;
+            const float *biasData = nullptr;
+            if (bias != nullptr)
+                biasData = (const float *)bias->data;
+
+            for (int ofs = stripeStart; ofs < stripeEnd; ++ofs)
+            {
+                const float* first = srcData + ofs * normSize;
+                float* dst_first = dstData + ofs * normSize;
+
+                float mean = 0;
+                float mean_square = 0;
+                for (int h = 0; h < normSize; ++h) {
+                    mean += first[h];
+                    mean_square += first[h] * first[h];
+                }
+                mean /= normSize;
+                mean_square = std::sqrt(mean_square / normSize - mean * mean + epsilon);
+                for (int h = 0; h < normSize; ++h) {
+                    if (biasData == nullptr) {
+                        dst_first[h] = (first[h] - mean) / mean_square * scaleData[h];
+                    } else {
+                        dst_first[h] = (first[h] - mean) / mean_square * scaleData[h] + biasData[h];
+                    }
+                }
+            }
+        }
+    };
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        if (inputs_arr.depth() == CV_16S)
+        {
+            forward_fallback(inputs_arr, outputs_arr, internals_arr);
+            return;
+        }
+
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+        const int nstripes = getNumThreads();
```

Comment:
> `const int nstripes = getNumThreads();`

This doesn't look as a reliable design.
This scheme assumes that all threads has the same speed and they are not interrupted.
It is not true for OS with preemptive execution (all widely used OS for the last 25+ years). Some cores may handle background tasks or interrupts.

Also it is not true at all for CPUs with big+little design (modern ARM, Intel CPUs with P+E cores).

`nstripes` should be based on subtask's "grain size" (subtask time >>> scheduling overhead) instead of number of available threads.

Some information is available here: https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Controlling_Chunking_os.html

Reply:
> Also it is not true at all for CPUs with big+little design (modern ARM, Intel CPUs with P+E cores).

And this is why I want to fix the openmp issue on macOS. I was doing benchmarking on vision transformers on onnxruntime, mnn and opencv dnn the other day, and found that both onnxruntime and mnn can run with 4 threads on my apple m1 by default, but opencv dnn uses all threads (8) instead.

I think somehow we should improve the multi-threading functionality of opencv, which detects big+little design and returns the numThreads of big cores if possible. Somehow `cv::Range` needs to support step as well in order to work with "grain size"...

Reply:
I am not sure what you are exactly asking for here. Basically every other `**Invoker` of other layers uses the same strategy. If we want to take care of big+little core CPUs, that is going to be another pull request I think.